scapy/pton_ntop.py

```diff
@@ -77,25 +77,53 @@ def inet_ntop(af, addr):
         except AttributeError:
             pass
 
-        # IPv6 addresses have 128bits (16 bytes)
-        if len(addr) != 16:
+        return _ipv6_bin_to_str(addr)
+    else:
+        raise Exception("Address family not supported yet")
+
+
+def _ipv6_bin_to_str(addr):
+    # IPv6 addresses have 128bits (16 bytes)
+    if len(addr) != 16:
+        raise Exception("Illegal syntax for IP address")
+    parts = []
+    for left in [0, 2, 4, 6, 8, 10, 12, 14]:
+        try:
+            value = struct.unpack("!H", addr[left:left + 2])[0]
+            hexstr = hex(value)[2:]
+        except TypeError:
             raise Exception("Illegal syntax for IP address")
-        parts = []
-        for left in [0, 2, 4, 6, 8, 10, 12, 14]:
-            try: 
-                value = struct.unpack("!H", addr[left:left+2])[0]
-                hexstr = hex(value)[2:]
-            except TypeError:
-                raise Exception("Illegal syntax for IP address")
-            parts.append(hexstr.lstrip("0").lower())
-        result = ":".join(parts)
-        while ":::" in result:
-            result = result.replace(":::", "::")
-        # Leaving out leading and trailing zeros is only allowed with ::
-        if result.endswith(":") and not result.endswith("::"):
-            result = result + "0"
-        if result.startswith(":") and not result.startswith("::"):
-            result = "0" + result
-        return result
+        parts.append(hexstr.lower())
+
+    address = ":".join(parts)
+
+    # Find all consecutive zero blocks
+    matches = re.findall('(?::|^)(0(?::0)+)(?::|$)', address)
```

Comment:
Using `re.findall()` produces a (useless) list, plus you lose the regexp match object (which has interesting attributes for what you want to do).
I think the code starting from here to the end of the function can be rewritten as:
```Python
    try:
        match = max(re.finditer('(?::|^)(0(?::0)+)(?::|$)', address),
                    key=lambda x: x.end() - x.start())
    except ValueError:
        return address
    return '%s::%s' % (address[:match.start()], address[match.end():])
```

I have not testes that code, however, so it might require some adjustments.